### PR TITLE
net: add track_caller to public APIs

### PR DESCRIPTION
**Pull request by hds**
## Motivation

When a user of tokio calls a function that panics when misused (e.g. creating
a tokio `TcpListener` from a std `TcpListener` without having an IO driver
enabled in the surrounding tokio runtime) then the user currently sees the
line number of the panic call inside tokio. It would be more informative for
the user to see the place where they called the panicking function.

It is still possible for the user to see the full stack trace by setting the
environment variable RUST_BACKLOG=1, so no useful information is
hidden.

This change is the 5th in a series towards closing #4413 (starting with #4772),
this change is for the net functions in the main tokio crate.

## Solution

Functions that may panic can be annotated with #[track_caller] so that
in the event of a panic, the function where the user called the
panicking function is shown instead of the file and line within Tokio
source.

This change adds #[track_caller] to all the non-unstable public net APIs
in the main tokio crate where the documentation describes how the
function may panic due to incorrect context or inputs.  Not all panic
cases can have #[track_caller] applied fully as the callstack passes
through a closure which isn't yet supported by the annotations (e.g. net
functions called from outside a tokio runtime).

Additionally, the documentation was updated to indicate additional cases
in which the public net functions may panic (the same as the io
functions).

Tests are included to cover each potentially panicking function.

Refs: #4413

### CODE REVIEW

File: tokio/src/net/tcp/listener.rs [hunk 1]
```diff
@@ -216,11 +216,13 @@ impl TcpListener {
     ///
     /// # Panics
     ///
-    /// This function panics if thread-local runtime is not set.
+    /// This function panics if there is no current reactor set, if the `rt`
+    /// feature flag is not enabled, or if thread-local runtime is not set.
```
Comment (Darksonn) [inline tokio/src/net/tcp/listener.rs:220]: Setting a current reactor or thread-local runtime is the same thing?

Comment (hds) [reply tokio/src/net/tcp/listener.rs:220]: OK, and is my understanding correct that this also covers needing `enable_io` on the runtime?

Does it make sense to keep the description of the case that the `rt` feature flag is not enabled? Or would it be better to leave the comment as it was (here and in all the other places I changed it)?

Comment (Darksonn) [reply tokio/src/net/tcp/listener.rs:220]: Ah, I see what you meant. You can just say that it panics if you are not inside a runtime with IO enabled. You can't be inside such a runtime if `rt` is not enabled, so we don't need to mention that.

Comment (hds) [reply tokio/src/net/tcp/listener.rs]: I've tried to write something that makes sense. Please let me know if it needs changing.

File: tokio/src/net/tcp/listener.rs [hunk 2]
```diff
@@ -216,11 +216,13 @@ impl TcpListener {
     ///
     /// # Panics
     ///
-    /// This function panics if thread-local runtime is not set.
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled or if the thread-local runtime is not set.
```
Comment (Darksonn) [inline tokio/src/net/tcp/listener.rs:220]:
```suggestion
    /// This function panics if it is not called from within a runtime with
    /// IO enabled.
```

Comment (Darksonn) [reply tokio/src/net/tcp/listener.rs:220]: @hds Did you have any thoughts on this doc change?

Comment (hds) [reply tokio/src/net/tcp/listener.rs:220]: @Darksonn I agree with them, it fits in with the rest of the docs. I'll make them in all the relevant spots in the next few days - I've been on holiday and have extended that to not opening my computer at all for a week. (-;

Comment (Darksonn) [reply tokio/src/net/tcp/listener.rs:220]: Alright. The next release is probably a month from now, so there's no rush.

Comment (hds) [reply tokio/src/net/tcp/listener.rs]: Fixed.

**Review by Darksonn — APPROVED** [review]
Thanks!

File: tokio/tests/net_panic.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+use tokio::net::{TcpListener, TcpStream};
+use tokio::runtime::{Builder, Runtime};
+
+mod support {
+    pub mod panic;
+}
+use support::panic::test_panic;
+
+#[test]
+fn udp_socket_from_std_panic_caller() -> Result<(), Box<dyn Error>> {
```
Comment (Darksonn) [inline tokio/tests/net_panic.rs:14]:
```suggestion
#[test]
#[cfg(not(target_os = "wasi"))]
fn udp_socket_from_std_panic_caller() -> Result<(), Box<dyn Error>> {
```

File: tokio/tests/net_panic.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+use tokio::runtime::{Builder, Runtime};
+
+mod support {
+    pub mod panic;
+}
+use support::panic::test_panic;
+
+#[test]
+#[cfg(not(target_os = "wasi"))]
+fn udp_socket_from_std_panic_caller() -> Result<(), Box<dyn Error>> {
```
Comment (Darksonn) [inline tokio/tests/net_panic.rs:15]: This needs to go on the entire file. WASM doesn't support `catch_unwind`. Sorry about that.

Comment (hds) [reply tokio/tests/net_panic.rs]: Thank you, I was just digging into the Wasm32-wasi PR to work out what I was doing wrong. That saved me a bunch of time!

**Pull request merged**